test/integration/test_controlplane.go

```diff
@@ -392,6 +397,186 @@ func TestControlPlaneEssentials(t *testing.T) {
+	dp, err := dataplaneClient.Create(GetCtx(), dp, metav1.CreateOptions{})
+	require.NoError(t, err)
+	cleaner.Add(dp)
+
+	createNamespace := func(t *testing.T, cl client.Client, cleaner *clusters.Cleaner, generateName string) *corev1.Namespace {
+		ns := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: generateName,
+			},
+		}
+		require.NoError(t, cl.Create(GetCtx(), ns))
+		cleaner.AddNamespace(ns)
+		return ns
+	}
+	nsA := createNamespace(t, GetClients().MgrClient, cleaner, "test-namespace-a")
+	nsB := createNamespace(t, GetClients().MgrClient, cleaner, "test-namespace-b")
+
+	cp := &operatorv1beta1.ControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    namespace.Name,
+			GenerateName: "cp-watchnamespaces-",
+		},
+		Spec: operatorv1beta1.ControlPlaneSpec{
+			ControlPlaneOptions: operatorv1beta1.ControlPlaneOptions{
+				Deployment: operatorv1beta1.ControlPlaneDeploymentOptions{
+					PodTemplateSpec: &corev1.PodTemplateSpec{
+						Spec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Name:  consts.ControlPlaneControllerContainerName,
+									Image: consts.DefaultControlPlaneImage,
+								},
+							},
+						},
+					},
+				},
+				DataPlane: lo.ToPtr(dp.Name),
+			},
+			WatchNamespaces: &operatorv1beta1.WatchNamespaces{
+				Type: operatorv1beta1.WatchNamespacesTypeList,
+				List: []string{
+					nsA.Name,
+					nsB.Name,
+				},
+			},
+		},
+	}
+
+	t.Log("deploying controlplane resource")
+	cp, err = controlplaneClient.Create(GetCtx(), cp, metav1.CreateOptions{})
+	require.NoError(t, err)
+	cleaner.Add(cp)
+
+	t.Log("verifying controlplane has a status condition indicating missing WatchNamespaceGrants")
+	require.Eventually(t,
+		testutils.ObjectPredicates(t, clients.MgrClient,
+			testutils.MatchCondition[*operatorv1beta1.ControlPlane](t).
+				Type(string(kcfgcontrolplane.ConditionTypeWatchNamespaceGrantValid)).
+				Status(metav1.ConditionFalse).
+				Reason(string(kcfgcontrolplane.ConditionReasonWatchNamespaceGrantInvalid)).
+				Predicate(),
+			testutils.MatchCondition[*operatorv1beta1.ControlPlane](t).
+				Type(string(kcfgdataplane.ReadyType)).
+				Status(metav1.ConditionFalse).
+				Predicate(),
+		).Match(cp),
+		testutils.ControlPlaneCondDeadline, 2*testutils.ControlPlaneCondTick,
+	)
+
+	watchNamespaceGrantForNamespace := func(t *testing.T, cl client.Client, cp *operatorv1beta1.ControlPlane, ns string) *operatorv1alpha1.WatchNamespaceGrant {
+		wng := &operatorv1alpha1.WatchNamespaceGrant{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: ns + "-refgrant-",
+				Namespace:    ns,
+			},
+			Spec: operatorv1alpha1.WatchNamespaceGrantSpec{
+				From: []operatorv1alpha1.WatchNamespaceGrantFrom{
+					{
+						Group:     operatorv1beta1.SchemeGroupVersion.Group,
+						Kind:      "ControlPlane",
+						Namespace: cp.Namespace,
+					},
+				},
+			},
+		}
+
+		require.NoError(t, cl.Create(GetCtx(), wng))
+		return wng
+	}
```

Comment:
Can we extract this into a helper?

Reply:
This is already extracted to a func 😅 but if you meant to place it outside of the test then 613e846c3a76983770d6d274bf2654d72b280181 does that.

Reply:
yep, that's what I meant 😄